kong/dns/client.lua

```diff
@@ -0,0 +1,783 @@
+  local time_str = ("%.3f %.3f"):format(start_time, query_time)
+
+  stats_set_count(self.stats, key, "query_last_time", time_str)
+
+  log(DEBUG, "r:query(", key, ") ans:", answers and #answers or "-", " t:", time_str)
+
+  if not answers then
+    stats_increment(self.stats, key, "query_fail_nameserver")
+    err = err or "unknown"
+    return nil, "DNS server error: " .. err .. ", Query Time: " .. time_str
+  end
+
+  answers = process_answers(self, name, qtype, answers)
+
+  stats_increment(self.stats, key, answers.errstr and "query_fail:" .. answers.errstr
+                                                   or "query_succ")
+
+  return answers, nil, answers.ttl
+end
+
+
+local function stale_update_task(premature, self, key, name, qtype, short_key)
+  if premature then
+    return
+  end
+
+  local answers = resolve_query(self, name, qtype)
+  if answers and (not answers.errcode or answers.errcode == NAME_ERROR_CODE) then
+    self.cache:set(key, { ttl = answers.ttl }, answers)
+    insert_last_type(self.cache, name, qtype)
+
+    -- simply invalidate it and let the search iteration choose the correct one
+    self.cache:delete(short_key)
+  end
+end
+
+
+local function start_stale_update_task(self, key, name, qtype, short_key)
+  stats_increment(self.stats, key, "stale")
+
+  local ok, err = timer_at(0, stale_update_task, self, key, name, qtype, short_key)
+  if not ok then
+    log(ALERT, "failed to start a timer to update stale DNS records: ", err)
+  end
+end
+
+
+local function resolve_name_type_callback(self, name, qtype, cache_only, short_key, tries)
+  local key = name .. ":" .. qtype
+
+  -- check if this key exists in the hosts file (it maybe evicted from cache)
+  local answers = self.hosts_cache[key]
+  if answers then
+    return answers, nil, answers.ttl
+  end
+
+  -- `:peek(stale=true)` verifies if the expired key remains in L2 shm, then
+  -- initiates an asynchronous background updating task to refresh it.
+  local ttl, _, answers = self.cache:peek(key, true)
+  if answers and ttl then
+    if not answers.expired then
+      answers.expire = now() + ttl
+      answers.expired = true
+      ttl = ttl + self.stale_ttl
+
+    else
+      ttl = ttl + (answers.expire - now())
+    end
+
+    -- trigger the update task by the upper caller every 60 seconds
+    ttl = math_min(ttl, 60)
+
+    if ttl > 0 then
+      log(DEBUG, "start stale update task ", key, " ttl:", ttl)
+
+      -- mlcache's internal lock mechanism ensures concurrent control
+      start_stale_update_task(self, key, name, qtype, short_key)
+      answers.ttl = ttl
+      return answers, nil, ttl
+    end
+  end
+
+  if cache_only then
+    return CACHE_ONLY_ANSWERS, nil, -1
+  end
+
+  local answers, err, ttl = resolve_query(self, name, qtype)
+  return answers, err, ttl
+end
+
+
+-- detect circular references in DNS CNAME or SRV records
+local function detect_recursion(resolved_names, key)
+  local detected = resolved_names[key]
```

Comment:
So I was thinking something like this (see the following comments too):

```suggestion
  if not resolved_names then
    return
  end
  local detected = resolved_names[key]
```